### ISSUE TITLE
Bug contourf

### ISSUE THREAD

**Issue by raphiclimate** [lukelbd/proplot#283]
### Description

I'm running into an error after upgrading the proplot version . The error occurs when using contourf, the error message is KeyError: 'red'

### Steps to reproduce

```python
# your code here
# we should be able to copy-paste this into python and exactly reproduce your bug

import proplot as plot
import matplotlib.pyplot as plt
import numpy as np

f, axs = plot.subplots(ncols=1, nrows=1) 

#generate random data
data=np.ones((180,360))
data[20:80,80:150]+=1

x=np.arange(0,360,1)
y=np.arange(0,180,1)

axs.contourf(x,y,data,cmap=plt.cm.coolwarm)

plt.show()
```
**Expected behavior**:  A simple contourf plot

**Actual behavior**: An error message occured and the plot was not produced

### Equivalent steps in matplotlib

Pretty sure its Proplot. With matplotlib this works
```python
import matplotlib.pyplot as plt
import numpy as np

f,axs=plt.subplots()

#generate random data
data=np.ones((180,360))
data[20:80,80:150]+=1
x=np.arange(0,360,1)
y=np.arange(0,180,1)

axs.contourf(x,y,data,cmap=plt.cm.coolwarm)
plt.show()
```
### Proplot version
Matplotlib: 3.4.1
Proplot: 0.9.1

**Comment by matthias-k** [comment]
I think I ran into the same bug in a slightly different setting, which might help to debug the issue:

```python
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import proplot as pplt

f, axs = pplt.subplots(
    nrows=1, ncols=1,
)

data = np.random.randn(2, 100)
axs[0].hist2d(
        data[0], data[1],
        cmap=plt.cm.gray_r
)
axs[0].arrow(-1, 0, 1, 0, color='blue')

axs[0].hist2d(
        data[0], data[1],
        cmap=plt.cm.gray_r
)
```

Here, the second hist2d function also throws the `'red'` error. Interestingly, the example works if I
* don't use `cmap=plt.cm.gray_r` with either of the two `hist2d` calls
* or remove the first `hist2d` call completely
* or use `plt.subplots` instead of `plt.subplots`
* or create the axes simply via `ax=plt.gca()`

So, to summarize, you really need to call `hist2d` twice, with a call to `arrow` in between, and all of that on an axes object that was created via `pplt.subplots`. The second `hist2d` call can also go to a different axes, it doesn't have to be the same as the first axes.

**Comment by matthias-k** [comment]
I did some more digging. I know that this code worked with an older version of proplot, probably around v6.4. With matplotlib 3.4.2, the bug is already present in version 0.7, and earlier versions of proplot seem to be incompatible with matplotlib 3.4.2. Therefore I installed matplotlib 3.3.4, and found that the bug is not yet present in 0.6.4, but present in 0.7.0.

To pinpoint the source of the bug further, I ran git bisect between 0.6.4 and 0.7.0. Unfortunately, there were a lot of commits with other bugs  (usually `'tuple' object has no attribute 'cmap'`), so git bisect could not exactly locate the commit that introduced the bug, but it claims it could reduce it to only two commits:

```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
2f87b04a5ac06ceeff355c60d6b3e0be984060be
3d8e7dd0ce16e1f9ac98b430899ecead94648959
We cannot bisect more!
bisect run cannot continue any more
```
So apparently, it's either 2f87b04a5ac06ceeff355c60d6b3e0be984060be or 3d8e7dd0ce16e1f9ac98b430899ecead94648959

Let me know if there is more that I can do to help!

**Comment by lukelbd** [comment]
Thanks for the report @raphiclimate and the help @matthias-k.

It looks like the segmentdata dictionary is for some mysterious reason cleaned out in the colormap instances stored under the `cm` module: `plt.cm.coolwarm._segmentdata` returns `{}`. For now, I think using the colormap names instead of these instances should work, e.g. `cmap='coolwarm'` or `cmap='grays'`.  But I'll look into this.

**Comment by lukelbd** [comment]
To be clear that's the source of the `KeyError` -- normally `segmentdata` contains red, blue, and green channel transitions (or hue, luminance, saturation for a `PerceptualColormap`). It looks like just *importing* proplot and nothing else breaks the colormaps stored under the `cm` module. Strange.

**Comment by matthias-k** [comment]
@lukelbd thanks for the quick response. Your workaround saved me a lot of work (I have a submission deadline coming up and was already thinking about setting up a new enviroment with the old proplot version just for adapting a few figures ;)).

An additional clarification: I guess it's not only the import. Otherwise I would expect that (in my example from above) already the first `hist2d` call fails, and not only the second one after the `arrow` call.

**Comment by lukelbd** [comment]
No problem. Thanks for your clarification -- after that I found the solution relatively quickly.

The `Colormap` constructor function converts input native matplotlib colormaps to proplot's colormap subclasses (i.e. `ListedColormap` to `DiscreteColormap`, `LinearSegmentedColormap` to `ContinuousColormap`). There was a bug in the conversion process that caused erasure of the original `segmentdata`. This happened the first time the native matplotlib colormap was passed through `Colormap` (i.e., the first `hist2d` call). To fix this, I simply make sure to copy the native colormap segment data dictionary when using it for the new colormap.

**Comment by raphiclimate** [comment]
Thanks @matthias-k for digging further and @lukelbd for addressing the issue so quickly!
